### ISSUE TITLE
Random musings on supervision

### ISSUE THREAD

**Issue by jkbbwr** [dabeaz/curio#138]
Rather than have curio manage supervision of tasks and subtasks, why not expose a way to link tasks. That way if I have a task that spawns a subtask I spawn the task and link myself to it. If that task dies an exception is raised in my task of the type `Subtask Death` and I can decide how to act on that.

**Comment by imrn** [comment]
Isn't it similar to task cancellation? When a task is cancelled, exception
is raised in the task. Seems as a good idea. Do you have a suggestion
for linking api?

I'm wondering if a pattern similar to cancellation be adapted to this.
Even further can this be generalized as an inter task exception
exchange?

**Comment by jkbbwr** [comment]
Maybe `.link(task: Task)` on a given Task. 
`.link` is a one directional link of child -> parent. If child throws an exception then we should be able to raise an exception inside the parent.

**Comment by dabeaz** [comment]
I don't think subtasks should report errors via exceptions to their parent task except explicitly through the `task.join()` method.   There are a whole bunch of problems.

1.  The only place curio can even raise an exception in a task is at a yield-point in an an async call.  So, it would only happen on statements involving `await`.
2.  A subtask could crash at any point in time meaning that the exception could possibly be raised anywhere.  This means that you'd have to check for it everywhere.   That would be a mess.
3. Related to (2), I see a whole host of problems that seem similar to issues surrounding things like Unix signal handling or interrupted system calls.   For example, if an operation fails due to a subtask death and you want to keep going, do you need to put retry loops on everything?  It could get really ugly, really fast.

In the big picture, I think it will be much more sane to have dead tasks handled in a very orderly manner.  Right now, the `join()` method already reports an exception if a task crashes.  If crashed tasks could additionally be reported on a queue, they could be monitored by a dedicated supervisor or some sort.  On the whole, I like that much better.

**Comment by imrn** [comment]
I'd probably revert to my original stance:

- Task management would NOT be the only thing I will
be dealing on an exception. I should be handling it
properly and providing required communication channels
covering all possible situations in the app.

- If an exception is left unhandled in a task, than that means:
I don't care about the task.

- So the idea of a mechanism for dealing with unhandled
exception is self defeating.

HOWEVER, I'm looking to this:

Doing something here and getting an exception
elsewhere is something interesting. Examples are:

- Cancellation: This is infact Task1 sending an exception
to Task2.
- Timeout: Task1 sends an future exception to itself.

They are something like intertask communication.
Although this is called "exception" (which does not
sound as a good thing) the receiver can receive
it at "Any Awaiting Point" of its execution. (Its for cancellation.)

Well I'm still not sure whether this is a good
thing or not. But this is definetely an emergent property
of asyncs.

This may be something that can be abused
or with some further generalization effort can it be
turned into a interesting expressive construct?

Any thoughts?

**Comment by njsmith** [comment]
@jkbbwr appears to be assuming that everyone here is familiar with how Erlang handles task death signaling, which might be a bit optimistic :-). Here's an excellent article on how [Erlang handles these kinds of things in general](http://ferd.ca/the-zen-of-erlang.html). IIUC, the way linking in particular works is:

* Links are bidirectional, and by default the lifetimes of the tasks are "linked" in the sense that if one dies, they both die. (And this is of course transitive, so if you think of links as making edges in a graph, then when one task dies, its whole connected component dies.) This kind of bidirectional linked-fates semantics seems handy for lots of common simple cases like, for example, the worker tasks in `curio.subprocess.Popen.communicate`, where you want parent cancellation to trigger child cancellation, and child crashes to propagate and cancel the other children.

* A given task can override this behavior, so that if one of its linked tasks dies, then instead of getting canceled it instead receives a message in its "mailbox". (All Erlang processes have a dedicated "mailbox", which is an incoming mostly-FIFO message queue.)

(Erlang also has "monitors", which are a way for a task to request to be notified when some arbitrary other task dies. This is different from links in that monitors are unidirectional, never kill the receiving process, and there can be multiple monitors watching the same task.)

So the obvious problem here is that Curio does not have the concept of signals, or task mailboxes :-). I've actually wondered independently whether we should generalize the current cancellation concept to something more like a "cancel signal" that has the default behavior of raising an exception, but where processes have the ability to redirect that to some sort of message queue. (Similar to how SIGTERM by default kills a Unix process, but the process can either mask it temporarily -- like `curio.defer_cancellation` -- or register an arbitrary handler -- curio doesn't have an analogy for this.) If we did have a more general "signal system" like this then Erlang-style links might make more sense?

Agreed with @dabeaz though that something more like Erlang's "monitors" with an explicit queue are definitely the thing to add first.

**Comment by njsmith** [comment]
Ah, here's the [Erlang docs on linking and monitors](http://erlang.org/doc/reference_manual/processes.html)

**Comment by imrn** [comment]
Note that, during cancellation your target is NOT a specific
queue, mailbox or similar, but the "whole task which
might be awaiting anywhere".

Your message/exception will surface whereever
the target task currently is. (Thanks to asyncs)

I admit that this is an amorph thing to
make something concrete. But it may lead
into some other thing which may be interesting.
May be not..

**Comment by njsmith** [comment]
Instead of (or in addition to) my continuing to scatter links around these threads, I just consolidated some of them on a wiki page: https://github.com/dabeaz/curio/wiki/Reading-list -- please add your own favorite articles as well :-)

**Comment by jkbbwr** [comment]
Because I want to try and grow an actor framework on top and if I get linking out of my scheduler then I get most of the actor framework for free

**Comment by dabeaz** [comment]
A followup comment.   I'll take a look at this linking stuff.  There could be some interesting things there.

My main concern:  I don't think that a task should report anything about its status (or termination) to another task except via an explicit `task.join()` operation.    The idea of having tasks linked and cancelled when a another task dies is slightly different issue.  Tasks can already be cancelled at any time in Curio.   So, one could certainly envision different ways of initiating that cancellation.

**Comment by jkbbwr** [comment]
Is there a way for a task to be aware of its own cancellation and act on it?

**Comment by imrn** [comment]
There will be an exception in the task.
You can handle it and do what ever you want.
As you can notice we're all talking about innovative
uses of such capability.

From documentation:
await Task.cancel(blocking=True): Cancels the task. This raises a curio.
CancelledError exception in the task which may choose to handle it in
order to perform cleanup actions.

**Comment by jkbbwr** [comment]
So currently if you cancel a task an exception is thrown inside it. Could you not do the same thing with a linked task?

e.g.

```
import curio

async def countdown(n):
    while n > 0:
        print('T-minus', n)
        await curio.sleep(1)
        n -= 1

async def sabotage(n):
    await curio.sleep(n/2)
    raise Exception("I died")

async def launch_manager():
    launchpad = await curio.spawn(countdown(10))
    spy = await curio.spawn(sabotage(10))
    spy.link(launchpad) # Spy reports failures upward
    await launchpad.join()
```

Then when `spy` raises an exception a `LinkedTaskFailure` or similar named exception can be thrown in the `launchpad`

**Comment by imrn** [comment]
You can do the linking with something like that:

```python
class KillerTask:
    def __init__(self, victimtask):
        self.victim = victimtask

    async def run():
        await sleep(5)
        await self.victim.cancel()

kt = KillerTask(victim)
spawn(kt.run())
```

**Comment by jkbbwr** [comment]
Thats only if the task dies normally, I am talking about exceptional failure also

**Comment by jkbbwr** [comment]
@dabeaz 
What if we were to do something like

```
async def always_on_failure(exc, coro, context):
    if exc is not None:
        await curio.spawn(countdown(), supervisor=always_on_failure)


async def countdown():
    for i in reversed(range(10)):
        if i == 3:
            if random.randint(0, 100) < 50:
                raise Exception("Rocket blew up :(")
        await curio.sleep(.1)
        print("T MINUS %s" % i)



curio.run(countdown(), with_supervisor=always_on_failure)
```

**Comment by imrn** [comment]
A link I wanted to share:
https://simonmar.github.io/posts/2017-01-24-asynchronous-exceptions.html

What interested me is the idea of inverting the exception handling flow.
That is: "Only handle the exception where you need to or care". Otherwise
let it go unhandled and exit your current thread/task/context/whatever..

It may be food for thought about task exceptions, cancellations or
previous idea of sending exceptions(messages or what_ever_you_call)
to tasks, etc..

**Comment by imrn** [comment]
And note that if there would be something like "exception as a message"
then handling AND continuing within the task would be an legal thing.
They can be mechanism of receiving external messages without polling.

**Comment by dabeaz** [comment]
Thanks for that link!  That was super interesting.   I must say, the whole handling of timeouts/cancellation might be one of the most interesting aspects of Curio right now.  I'm still not sure if it's completely insane or awesome.  There's a fine line between the two perhaps.

**Comment by imrn** [comment]
Such an approach may bring a radically different degree of freedom
for flow control, which may also lead to loss of structure on improper use.
This degree of freedom may not mean a lot for sync code. But I think
it is relevant for asyncs. I suggest experimenting with the mechanism in some
form (without comprimising current structure and without much investment).

**Comment by dabeaz** [comment]
Reading more on those Haskell papers and asynchronous exceptions, I can't quite decide if Curio is on the right path or a path to later madness.     Maybe this will be my excuse to learn Haskell.

**Comment by imrn** [comment]
I never meant a serious effort for it. As I also stated previously, the
question in my mind is just "Can we make cancellation/timeout
mechanism generic so that such exceptions can also be sent on demand?".

All the noise I was making is just this, knowing that you'll be spending
sometime at that neighbourhood last month.

I dont' think knowledge of haskell or java is required for this.
(Article says java once had it but later discarded it.)

Anyway, if it seems costly for Curio (based on your feedback) we can easily
forget it. But I feel Curio is close if generalization of cancellation/timeout/...
is possible..

**Comment by dabeaz** [comment]
Personally, I feel like the behavior of timeouts and cancellation is on a pretty good track right now.  There are DEFINITELY some tricky situations, but with timeout nesting and the cancellation control features that Nathan introduced, it seems like there is a certain about of sanity in the implementation.  I'm sort of surprised how similar it seems to be to some of that Haskell work you linked. For example, the paper describing block and unblock directives to control interrupt delivery and other things. 

Whether or not Curio expands to allow any exception to be delivered asynchronously, I just don't know.  Implementation wise, it might not be much.  Understanding the semantics of it at an application level could be a whole different story.  I just don't know.   In any event, it's still all very interesting.

**Comment by imrn** [comment]
How about "await task.raise(excp)"?
Just like "await task.cancel()".

**Comment by njsmith** [comment]
Python only has one actually asynchronous exception, and that's
KeyboardInterrupt, which actually can arrive between any bytecodes. Curio's
current design adds two restrictions: cancellations can only be delivered
when a task is yielded (really this is a Python restriction), and the task
being interrupted has to be in a state where it can unwind whatever it was
blocking for without side effects (so e.g. if it was about to take a lock
that's fine but if it just took a lock then we have to wait until the next
yield point). These both seem like good restrictions to me :-). That
article's really interesting, but it makes sense that a functional language
with async cancellation as a runtime primitive ends up with a somewhat
different design than a procedural language without that primitive.

If you're interested in reading about analogous designs, pthread
cancellation is extremely similar to curio's, though it turns out to be
much more annoying to deal with things at the C level. The musl developer's
blog at ewontfix.com has some interesting articles on this.

**Comment by njsmith** [comment]
Regarding task.raise(exc), what happens if two tasks call this with
different exceptions during the same event loop tick? (We can only deliver
one exception at a time.)

(Also raise can't be a method name in Python, because it's a keyword.)

**Comment by jkbbwr** [comment]
I mean some ideas of an API I have been playing with.

```
async def child():
    await sleep(5)
    raise Exception("woops")

async def parent():
    task = await spawn(child())
    await link(task)
    task, exception = await monitor()
```

Where `monitor` will just schedule the parent until there are no linked tasks running OR one of those linked tasks has failed with an exception.

Maybe this could be handled in userland like how `curio.wait` works
